src/Microsoft.Health.Fhir.Shared.Api/Controllers/ImportController.cs

```diff
@@ -79,9 +93,80 @@ public class ImportController : Controller
             _urlResolver = urlResolver;
             _features = features.Value;
             _mediator = mediator;
+            _resourceWrapperFactory = EnsureArg.IsNotNull(resourceWrapperFactory, nameof(resourceWrapperFactory));
             _logger = logger;
         }
 
+        [HttpPost]
+        [Route(KnownRoutes.ImportBundle)]
+        [AuditEventType(AuditEventSubType.ImportBundle)]
+        public async Task<IActionResult> ImportBundle()
+        {
+            var sw = Stopwatch.StartNew();
+            var startDate = DateTime.UtcNow;
+            var resources = new List<ImportResource>();
+            var keys = new HashSet<ResourceKey>();
+            var fhirParser = new FhirJsonParser();
+            var importParser = new ImportResourceParser(fhirParser, _resourceWrapperFactory);
+            var index = 0L;
+            using var reader = new StreamReader(Request.Body, Encoding.UTF8);
+            try
+            {
+                if (Request.ContentType != null && Request.ContentType.Contains("application/fhir+ndjson", StringComparison.OrdinalIgnoreCase))
+                {
+                    var line = await reader.ReadLineAsync();
+                    while (line != null)
+                    {
+                        var importResource = importParser.Parse(index, 0, 0, line, ImportMode.IncrementalLoad);
+                        var key = importResource.ResourceWrapper.ToResourceKey(true);
+                        if (!keys.Add(key))
+                        {
+                            throw new RequestNotValidException(string.Format(Resources.ResourcesMustBeUnique, key));
+                        }
+
+                        resources.Add(importResource);
+                        line = await reader.ReadLineAsync();
+                        index++;
+                    }
+                }
+                else
+                {
+                    var str = await reader.ReadToEndAsync();
+                    var resource = await fhirParser.ParseAsync(str);
+                    var bundle = resource.ToResourceElement().ToPoco<Bundle>();
```

Comment:
This bypasses the MVC InputFormatters - https://github.com/microsoft/fhir-server/blob/main/src/Microsoft.Health.Fhir.Shared.Api/Features/Formatters/FhirJsonInputFormatter.cs. 
I think this should be two different Action methods that delineate by media-type instead of trying to parse everything

Reply:
If I understand correctly, it would require 2 endpoints, something like: $import/ndjson and $import/batch. I am not a fan of this idea. I will try implementing checks for stream size you mentioned below.

Reply:
MVC natively supports routing to the correct action based on many factors such as URL, method, headers etc. The url can be the same, I'm more referring to something like https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.consumesattribute?view=aspnetcore-8.0, where for the same action, one can consume fhir+json, while the other consumes ndjson

Reply:
Main consumption is ndjson. We would need to put restrictions on size there. Will default MVC restrictions work? Will they limit content size to something that our service will not choke on?